src/Optimization/hiopKKTLinSysSparseCondensed.hpp

```diff
@@ -0,0 +1,281 @@
+
+  void times_diag_times_mat_numeric(const hiopVector& diag,
+                                    const hiopMatrixSparseCSRStorage& Y,
+                                    hiopMatrixSparseCSRStorage& M);
+
+public:
+
+  inline index_type* irowptr() 
+  {
+    return irowptr_;
+  }
+  inline index_type* jcolind()
+  {
+    return jcolind_;
+  }
+  inline double* values()
+  {
+    return values_;
+  }
+
+  inline index_type* irowptr() const
+  {
+    return irowptr_;
+  }
+  inline index_type* jcolind() const
+  {
+    return jcolind_;
+  }
+  inline double* values() const
+  {
+    return values_;
+  }
+  inline size_type m() const
+  {
+    return nrows_;
+  }
+  inline size_type n() const
+  {
+    return ncols_;
+  }
+  inline size_type nnz() const
+  {
+    return nnz_;
+  }
+  void print(FILE* file,
+             const char* msg=NULL,
+             int maxRows=-1,
+             int maxCols=-1,
+             int rank=-1) const;
+protected:
+  void alloc();
+  void dealloc();
+protected:
+  size_type nrows_;
+  size_type ncols_;
+  size_type nnz_;
+  index_type* irowptr_;
+  index_type* jcolind_;
+  double* values_;
+};
+
+/**
+ * Solves a sparse KKT linear system by exploiting the sparse structure, namely reduces 
+ * the so-called XDYcYd KKT system 
+ * [  H  +  Dx    0    Jd^T ] [ dx]   [ rx_tilde ]
+ * [    0         Dd   -I   ] [ dd] = [ rd_tilde ]
+ * [    Jd       -I     0   ] [dyd]   [   ryd    ]
+ * into the condensed KKT system
+ * (H+Dx+Jd^T*Dd*Jd)dx = rx_tilde + Jd^T*Dd*ryd + Jd^T*rd_tilde
+ * dd = Jd*dx - ryd
+ * dyd = Dd*dd - rd_tilde = Dd*Jd*dx - Dd*ryd - rd_tilde
+
+ * Here Jd is sparse Jacobians for inequalities, H is a sparse Hessian matrix, Dx is 
+ * log-barrier diagonal corresponding to x variables, Dd is the log-barrier diagonal 
+ * corresponding to the inequality slacks, and I is the identity matrix. 
+ *
+ * @note: the NLP is assumed to have no equality constraints (or have been relaxed to 
+ * two-sided inequality constraints).
+ *
+ *
+ * Dual regularization may be not enforced as it requires repeated divisions that are 
+ * to round-off error accumulation, but when it is enforced, the regularized XDYcYd KKT 
+ * system
+ * [  H+Dx+delta_wx*I         0         Jd^T     ] [ dx]   [ rx_tilde ]
+ * [          0         Dd+delta_wd*I   -I       ] [ dd] = [ rd_tilde ]
+ * [          Jd             -I         -delta_cd] [dyd]   [   ryd    ]
+ * is solved as:
+ * (notation) Dd2 = [ (Dd+delta_wd*I)^{-1} + delta_cd*I ]^{-1}
+ *
+ * dd = Jd*dx - delta_cd*dyd - ryd
+ *
+ * From (Dd+delta_wd*I)*dd - dyd = rd_tilde one can write
+ *   ->   (Dd+delta_wd*I)*(Jd*dx - delta_cd*dyd - ryd) - dyd = rd_tilde
+ *   ->   (I+ delta_cd*(Dd+delta_wd*I)) dyd = (Dd+delta_wd*I)*(Jd*dx - ryd) - rd_tilde 
+ * dyd = (I+ delta_cd*(Dd+delta_wd*I))^{-1} [ (Dd+delta_wd*I)*(Jd*dx - ryd) - rd_tilde ]
+ *
+ * (H+Dx+delta_wx*I + Jd^T * Dd2 * Jd) dx = rx_tilde + Jd^T*Dd2*(Dd+delta_wd*I)*ryd +  Jd^T*Dd2*rd_tilde
```

Comment:
I think 
```
Dd2 = [ I + delta_cd*(Dd+delta_wd*I)]^{-1}
and 
(H+Dx+delta_wx*I + Jd^T * Dd2 * (Dd+delta_wd*I) * Jd) dx = rx_tilde + Jd^T*Dd2*(Dd+delta_wd*I)*ryd + Jd^T*Dd2*rd_tilde
```

Reply:
it appears to me that they are equivalent, but your formulation may be better suited for computation (has smaller numbers to invert) and also has a negligible cost overhead...